sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala

```diff
@@ -462,27 +507,139 @@ class TypeCoercionSuite extends AnalysisTest {
       ArrayType(DoubleType, containsNull = false),
       Some(ArrayType(DoubleType, containsNull = true)))
     widenTestWithStringPromotion(
-      ArrayType(TimestampType, containsNull = false),
-      ArrayType(StringType, containsNull = true),
+      ArrayType(ArrayType(IntegerType), containsNull = true),
+      ArrayType(ArrayType(LongType), containsNull = false),
+      Some(ArrayType(ArrayType(LongType), containsNull = true)))
```

Comment:
Do we really wanna do this? Is there any other database have this behavior?

I think for complex type, we should ignore the nullable difference, but I'm not sure if we should do type coercion inside complex type.

Reply:
@mgaido91, thoughts on this?

It's definitely possible for us to revert back to behavior where we don't do IntegerType-to-LongType, xType-to-StringType, etc. promotion inside complex types, which was how a previous form of this PR handled it.

Reply:
I think this is more coherent with the behavior in other parts and this is the behavior I would expect. But I think that we should follow @gatorsmile's suggestion and check Hive's behavior first.